File: 2685-CounttheNumberofCompleteComponents/2685-CounttheNumberofCompleteComponents.py
class Solution:
    def countCompleteComponents(self, n: int, edges: List[List[int]]) -> int:
        graph = convert_to_graph(edges, n)
        visited = set()
        count = 0
        
        # Traverse all nodes in the graph
        for node in range(n):
            if node not in visited:
                component_nodes = set()
                if self.dfs(graph, node, visited, component_nodes):
                    # Check if the component forms a complete graph
                    if self.is_complete_graph(graph, component_nodes):
                        count += 1
        return count

    def dfs(self, graph, node, visited, component_nodes):
        visited.add(node)
        component_nodes.add(node)
        
        for neighbor in graph[node]:
            if neighbor not in visited:
                self.dfs(graph, neighbor, visited, component_nodes)
                
        return True
    
    def is_complete_graph(self, graph, component_nodes):
        # In a complete graph, every node has exactly len(component_nodes) - 1 neighbors
        component_size = len(component_nodes)
        for node in component_nodes:
            if len(graph[node]) != component_size - 1:
                return False
        return True

def convert_to_graph(edges, n):
    graph = {i: [] for i in range(n)}
    for a, b in edges:
        graph[a].append(b)
        graph[b].append(a)
    return graph

File: 2685-CounttheNumberofCompleteComponents/2685-CounttheNumberofCompleteComponents.js
/**
 * @param {number} n
 * @param {number[][]} edges
 * @return {number}
 */
var countCompleteComponents = function(n, edges) {
    //create a hash map with node a key and connection nodes as values
    let graph = createGraph(n, edges);
    let count = 0;
    let visited = new Set();
    

    // loop through the map
    Object.keys(graph).forEach((node, values) => {
        let componentNodes = new Set();
        const numNode = Number(node);
        // check if the key has not been visited
        if(!visited.has(numNode)){
            dfs(graph, numNode, visited, componentNodes);

            if(isACompleteComponet(graph, componentNodes) == true){
                count++;
            }
        }
    })
        // check if the node is in a complete component increase the count
          //dfs to explore all the nodes in the component 
    
    function dfs(graph, node, visited, componentNodes){
        if(visited.has(node)) return false
        
        visited.add(node);
        componentNodes.add(node);

        for(let neighbor of graph[node]){
            dfs(graph, neighbor, visited, componentNodes);
        }
        return true
    }

    function isACompleteComponet(graph, componentNodes){
        for(let node of componentNodes){
            if(graph[node].length != componentNodes.size - 1) return false;
        }
        return true;
    }

    function createGraph(n, edges) {
        let obj = {};
        //initialize the graph with the keys
        for(let i = 0; i < n; i++){
            obj[i] = [];
        }
        // iterate through each pair and add the connections to the graph
        for(let [a,b] of edges){
            obj[a].push(b);
            obj[b].push(a);
        }
        return obj
    }

    return count;
};
